crates/bevy_winit/src/accessibility.rs

```diff
@@ -0,0 +1,195 @@
+use std::{
+    collections::VecDeque,
+    num::NonZeroU128,
+    sync::{Arc, Mutex},
+};
+
+use accesskit::{ActionHandler, ActionRequest, Node, NodeId, Role, TreeUpdate};
+use accesskit_winit::Adapter;
+use bevy_app::{App, CoreStage, Plugin};
+use bevy_derive::{Deref, DerefMut};
+use bevy_ecs::{
+    prelude::{Component, Entity, EventReader, EventWriter},
+    query::{Changed, With},
+    system::{NonSend, NonSendMut, Query, RemovedComponents, Res, ResMut, Resource},
+};
+use bevy_utils::{default, HashMap};
+use bevy_window::{WindowClosed, WindowFocused, WindowId};
+
+#[derive(Component, Clone, Default, Deref, DerefMut)]
+pub struct AccessibilityNode(pub Node);
+
+impl From<Node> for AccessibilityNode {
+    fn from(node: Node) -> Self {
+        Self(node)
+    }
+}
+
+#[derive(Default, Deref, DerefMut)]
+pub struct Adapters(pub HashMap<WindowId, Adapter>);
+
+impl Adapters {
+    pub fn get_primary_adapter(&self) -> Option<&Adapter> {
+        self.get(&WindowId::primary())
+    }
+}
+
+#[derive(Resource, Default, Deref, DerefMut)]
+pub struct Handlers(pub HashMap<WindowId, WinitActionHandler>);
+
+#[derive(Clone, Default, Deref, DerefMut)]
+pub struct WinitActionHandler(pub Arc<Mutex<VecDeque<ActionRequest>>>);
+
+impl ActionHandler for WinitActionHandler {
+    fn do_action(&self, request: ActionRequest) {
+        println!("Pushing {:?}", request);
+        let mut requests = self.0.lock().unwrap();
+        requests.push_back(request);
+    }
+}
+
+pub trait AccessKitEntityExt {
+    fn from_node_id(id: &NodeId) -> Entity {
+        Entity::from_bits((id.0.get() - 1) as u64)
+    }
+
+    fn to_node_id(&self) -> NodeId;
+}
+
+impl AccessKitEntityExt for Entity {
+    fn to_node_id(&self) -> NodeId {
+        let id = NonZeroU128::new((self.to_bits() + 1) as u128);
+        NodeId(id.unwrap())
+    }
+}
+
+#[derive(Resource, Default, Deref, DerefMut)]
+pub struct LastFocus(Option<NodeId>);
+
+impl LastFocus {
+    pub fn from_entity(&mut self, entity: Option<Entity>) {
+        **self = entity.map(|v| v.to_node_id());
+    }
+
+    pub fn entity(&self) -> Option<Entity> {
+        self.0.as_ref().map(<Entity>::from_node_id)
+    }
+}
+
+fn handle_focus(
+    last_focus: Res<LastFocus>,
+    adapters: NonSend<Adapters>,
```

Comment:
Does this need to be `NonSend`? This is going to run up against #6552 eventually. If we can make it `Send`, it'd be for the best long term.

Reply:
Here are the errors I get when trying to drop the `NonSend` requirement. Not sure if there's anything I can do about them, but please let me know if there's some other way of storing this:

error[E0277]: `std::cell::Cell<bool>` cannot be shared between threads safely                                                                                                 
  --> crates\bevy_winit\src\accessibility.rs:28:10                                                                                                                            
   |                                                                                                                                                                          
28 | #[derive(Resource, Default, Deref, DerefMut)]                                                                                                                            
   |          ^^^^^^^^ `std::cell::Cell<bool>` cannot be shared between threads safely                                                                                        
   |                                                                                                                                                                          
   = help: within `accesskit_windows::subclass::SubclassImpl`, the trait `Sync` is not implemented for `std::cell::Cell<bool>`                                                
   = note: required because it appears within the type `accesskit_windows::subclass::SubclassImpl`                                                                            
   = note: required for `Unique<accesskit_windows::subclass::SubclassImpl>` to implement `Sync`                                                                               
   = note: required because it appears within the type `Box<accesskit_windows::subclass::SubclassImpl>`                                                                       
   = note: required because it appears within the type `accesskit_windows::subclass::SubclassingAdapter`                                                                      
   = note: required because it appears within the type `accesskit_winit::platform_impl::platform::Adapter`                                                                    
   = note: required because it appears within the type `Adapter`                                                                                                              
   = note: required because it appears within the type `(bevy_window::WindowId, Adapter)`                                                                                     
   = note: required for `RawTable<(bevy_window::WindowId, Adapter)>` to implement `Sync`                                                                                      
   = note: required because it appears within the type `bevy_utils::hashbrown::HashMap<bevy_window::WindowId, Adapter>`                                                       
note: required because it appears within the type `Adapters`                                                                                                                  
  --> crates\bevy_winit\src\accessibility.rs:29:12                                                                                                                            
   |                                                                                                                                                                          
29 | pub struct Adapters(pub HashMap<WindowId, Adapter>);                                                                                                                     
   |            ^^^^^^^^                                                                                                                                                      
   = help: see issue #48214                                                                                                                                                   
   = note: this error originates in the derive macro `Resource` (in Nightly builds, run with -Z macro-backtrace for more info)                                                
                                                                                                                                                                              
error[E0277]: `std::cell::Cell<Option<Box<(dyn FnOnce() -> accesskit_windows::adapter::Adapter + 'static)>>>` cannot be shared between threads safely                         
  --> crates\bevy_winit\src\accessibility.rs:28:10                                                                                                                            
   |                                                                                                                                                                          
28 | #[derive(Resource, Default, Deref, DerefMut)]                                                                                                                            
   |          ^^^^^^^^ `std::cell::Cell<Option<Box<(dyn FnOnce() -> accesskit_windows::adapter::Adapter + 'static)>>>` cannot be shared between threads safely                
   |                                                                                                                                                                          
   = help: within `accesskit_windows::subclass::SubclassImpl`, the trait `Sync` is not implemented for `std::cell::Cell<Option<Box<(dyn FnOnce() -> accesskit_windows::adapter::Adapter + 'static)>>>`                                                                                                                                                      
   = note: required because it appears within the type `once_cell::unsync::Lazy<accesskit_windows::adapter::Adapter, Box<(dyn FnOnce() -> accesskit_windows::adapter::Adapter + 'static)>>`                                                                                                                                                                 
   = note: required because it appears within the type `accesskit_windows::subclass::SubclassImpl`                                                                            
   = note: required for `Unique<accesskit_windows::subclass::SubclassImpl>` to implement `Sync`                                                                               
   = note: required because it appears within the type `Box<accesskit_windows::subclass::SubclassImpl>`                                                                       
   = note: required because it appears within the type `accesskit_windows::subclass::SubclassingAdapter`                                                                      
   = note: required because it appears within the type `accesskit_winit::platform_impl::platform::Adapter`                                                                    
   = note: required because it appears within the type `Adapter`                                                                                                              
   = note: required because it appears within the type `(bevy_window::WindowId, Adapter)`                                                                                     
   = note: required for `RawTable<(bevy_window::WindowId, Adapter)>` to implement `Sync`                                                                                      
   = note: required because it appears within the type `bevy_utils::hashbrown::HashMap<bevy_window::WindowId, Adapter>`                                                       
note: required because it appears within the type `Adapters`                                                                                                                  
  --> crates\bevy_winit\src\accessibility.rs:29:12                                                                                                                            
   |                                                                                                                                                                          
29 | pub struct Adapters(pub HashMap<WindowId, Adapter>);                                                                                                                     
   |            ^^^^^^^^                                                                                                                                                      
   = help: see issue #48214                                                                                                                                                   
   = note: this error originates in the derive macro `Resource` (in Nightly builds, run with -Z macro-backtrace for more info)                                                
                                                                                                                                                                              
error[E0277]: `UnsafeCell<Option<accesskit_windows::adapter::Adapter>>` cannot be shared between threads safely                                                               
  --> crates\bevy_winit\src\accessibility.rs:28:10                                                                                                                            
   |                                                                                                                                                                          
28 | #[derive(Resource, Default, Deref, DerefMut)]                                                                                                                            
   |          ^^^^^^^^ `UnsafeCell<Option<accesskit_windows::adapter::Adapter>>` cannot be shared between threads safely                                                      
   |                                                                                                                                                                          
   = help: within `accesskit_windows::subclass::SubclassImpl`, the trait `Sync` is not implemented for `UnsafeCell<Option<accesskit_windows::adapter::Adapter>>`              
   = note: required because it appears within the type `once_cell::unsync::OnceCell<accesskit_windows::adapter::Adapter>`                                                     
   = note: required because it appears within the type `once_cell::unsync::Lazy<accesskit_windows::adapter::Adapter, Box<(dyn FnOnce() -> accesskit_windows::adapter::Adapter + 'static)>>`                                                                                                                                                                 
   = note: required because it appears within the type `accesskit_windows::subclass::SubclassImpl`                                                                            
   = note: required for `Unique<accesskit_windows::subclass::SubclassImpl>` to implement `Sync`                                                                               
   = note: required because it appears within the type `Box<accesskit_windows::subclass::SubclassImpl>`                                                                       
   = note: required because it appears within the type `accesskit_windows::subclass::SubclassingAdapter`                                                                      
   = note: required because it appears within the type `accesskit_winit::platform_impl::platform::Adapter`                                                                    
   = note: required because it appears within the type `Adapter`                                                                                                              
   = note: required because it appears within the type `(bevy_window::WindowId, Adapter)`                                                                                     
   = note: required for `RawTable<(bevy_window::WindowId, Adapter)>` to implement `Sync`                                                                                      
   = note: required because it appears within the type `bevy_utils::hashbrown::HashMap<bevy_window::WindowId, Adapter>`                                                       
note: required because it appears within the type `Adapters`                                                                                                                  
  --> crates\bevy_winit\src\accessibility.rs:29:12                                                                                                                            
   |                                                                                                                                                                          
29 | pub struct Adapters(pub HashMap<WindowId, Adapter>);                                                                                                                     
   |            ^^^^^^^^                                                                                                                                                      
   = help: see issue #48214                                                                                                                                                   
   = note: this error originates in the derive macro `Resource` (in Nightly builds, run with -Z macro-backtrace for more info)                                                
                                                                                                                                                                              
error[E0277]: `(dyn FnOnce() -> accesskit_windows::adapter::Adapter + 'static)` cannot be sent between threads safely                                                         
  --> crates\bevy_winit\src\accessibility.rs:28:10                                                                                                                            
   |                                                                                                                                                                          
28 | #[derive(Resource, Default, Deref, DerefMut)]                                                                                                                            
   |          ^^^^^^^^ `(dyn FnOnce() -> accesskit_windows::adapter::Adapter + 'static)` cannot be sent between threads safely                                                
   |                                                                                                                                                                          
   = help: the trait `Send` is not implemented for `(dyn FnOnce() -> accesskit_windows::adapter::Adapter + 'static)`                                                          
   = note: required for `Unique<(dyn FnOnce() -> accesskit_windows::adapter::Adapter + 'static)>` to implement `Send`                                                         
   = note: required because it appears within the type `Box<(dyn FnOnce() -> accesskit_windows::adapter::Adapter + 'static)>`                                                 
   = note: required because it appears within the type `Option<Box<(dyn FnOnce() -> accesskit_windows::adapter::Adapter + 'static)>>`                                         
   = note: required for `std::cell::Cell<Option<Box<(dyn FnOnce() -> accesskit_windows::adapter::Adapter + 'static)>>>` to implement `Send`                                   
   = note: required because it appears within the type `once_cell::unsync::Lazy<accesskit_windows::adapter::Adapter, Box<(dyn FnOnce() -> accesskit_windows::adapter::Adapter + 'static)>>`                                                                                                                                                                 
   = note: required because it appears within the type `accesskit_windows::subclass::SubclassImpl`                                                                            
   = note: required for `Unique<accesskit_windows::subclass::SubclassImpl>` to implement `Send`                                                                               
   = note: required because it appears within the type `Box<accesskit_windows::subclass::SubclassImpl>`                                                                       
   = note: required because it appears within the type `accesskit_windows::subclass::SubclassingAdapter`                                                                      
   = note: required because it appears within the type `accesskit_winit::platform_impl::platform::Adapter`                                                                    
   = note: required because it appears within the type `Adapter`                                                                                                              
   = note: required because it appears within the type `(bevy_window::WindowId, Adapter)`                                                                                     
   = note: required for `RawTable<(bevy_window::WindowId, Adapter)>` to implement `Send`                                                                                      
   = note: required because it appears within the type `bevy_utils::hashbrown::HashMap<bevy_window::WindowId, Adapter>`                                                       
note: required because it appears within the type `Adapters`                                                                                                                  
  --> crates\bevy_winit\src\accessibility.rs:29:12                                                                                                                            
   |                                                                                                                                                                          
29 | pub struct Adapters(pub HashMap<WindowId, Adapter>);                                                                                                                     
   |            ^^^^^^^^                                                                                                                                                      
   = help: see issue #48214                                                                                                                                                   
   = note: this error originates in the derive macro `Resource` (in Nightly builds, run with -Z macro-backtrace for more info)

Reply:
It's a known limitation that the AccessKit adapters don't implement `Send`, and one that I don't plan on changing in the near future. For me, the question is whether arbitrary Bevy systems, like the ones in the Bevy UI accessibility plugin, really need to access the adapter. Maybe the solution is to add one or more additional commands, like the one that the Bevy UI accessibility plugin already uses to add or update AccessKit nodes.